### DISCUSSION TITLE
Module questions

### DISCUSSION THREAD

**Discussion by toyota-corolla0** [fedimint/fedimint#3341] [Q&A] [answered]
- Are inputs and outputs verified in a single threaded manner in the same order by all guardians?

- I have an in-memory data structure that I would like to mutably access in ServerModule::process_output. If the answer to my first question is no, then I need to rethink some stuff. If yes, how should I store that data structure.

Thanks for the help!!!

**Comment by dpc** [comment, marked as answer]
> Are inputs and outputs verified in a single threaded manner in the same order by all guardians?

The Federation needs to reach a consensus, so there has to be an ordering of how things are being processed. Anything else would be a serious bug.

I would need to check if we call `process_output` single-threaded in order, but I'd guess that yes.

>      If yes, how should I store that data structure.

In a database. Serialize using `Encodable` and store. See how other modules are storing their data. You might want to decompose it into individually update-able parts for performance reasons.

**Reply by toyota-corolla0** [reply]
Thanks for reply!

The data structure is a type of [orderbook](https://github.com/toyota-corolla0/rust_pie_ob) that could grow to be large. It is completely generated from orders in the database. I could regenerate it in every ```process_output``` but I believe that would be too slow. 

I came up with an idea to re implement the orderbook using ```ModuleDatabaseTransaction::find_by_prefix_sorted_descending``` so I will try that.

**Reply by dpc** [reply]
key-value stores can express any data-structure, so it's definitely possible.

**Reply by elsirion** [reply]
I'd avoid keeping mutable data outside the DB since you will not notice if e.g. the changes you made in `process_output` should be rolled back due to the transaction the output was part of being invalid. If all your mutable state is in the DB the database transaction will be rolled back and you don't have to worry about it.
